### PR TITLE
Explicit header whitelist in configuration

### LINKED ISSUE #317: 🚀 Configurable :request_headers / APPSIGNAL_REQUEST_HEADERS
We currently have a whitelist for request headers, which we'd like to make configurable.

https://github.com/appsignal/appsignal-elixir/blob/1f37ef36811089871db4cc0f773702426625212f/lib/appsignal/plug.ex#L157-L167

## The quick solution

The quickest to implement would be to [move the whitelist to the configuration](https://github.com/appsignal/appsignal-elixir/commit/10e694f98fb900cdb7b0a5eea93a41de16276b84). However, that makes it difficult for our users to add or remove a single header. Because `Appsignal.Config` isn't loaded int he host application's `config/appsignal.exs`, something like this won't work:

``` elixir
config :appsignal, :config,
  name: "appsignal-phoenix-example",
  push_api_key: "00000000-0000-0000-0000-000000000000",
  request_headers: Appsignal.Config.default_request_headers -- ~w(user-agent), # <- This won't work.
  env: Mix.env
````

To remove a single header from the config, one would have to copy the whole list into their configuration, and remove the one they don't want manually. 

``` elixir
config :appsignal, :config,
  name: "appsignal-phoenix-example",
  push_api_key: "00000000-0000-0000-0000-000000000000",
  request_headers: ~w(accept accept-charset accept-encoding accept-language cache-control connection content-length from negotiate pragma referer range auth-type gateway-interface path-translated remote-host remote-ident remote-user remote-addr request-method server-name server-port server-protocol request-uri path-info client-ip range x-request-start x-queue-start x-queue-time x-heroku-queue-wait-time x-application-start x-forwarded-for x-real-ip)
  env: Mix.env
```

## The explicit solution

Instead of having two places the configuration gets loaded from, the most explicit solution is to write the environment list to the configuration by default. Whenever a user uses the installer (`mix appsignal.install`), the generated config file will include the list of request headers. This will make sure there's one place where the request headers are listed.

We'll have to decide which headers we want to pick up by default, which will be a shorter list than we do now as users can now add their own request headers to the configuration.

``` elixir
config :appsignal, :config,
  name: "appsignal-phoenix-example",
  push_api_key: "00000000-0000-0000-0000-000000000000",
  request_headers: ~w(accept accept-charset accept-encoding accept-language cache-control connection content-length user-agent from negotiate pragma referer range),
  env: Mix.env
```

### Upgrading

For users with "old" configuration files that don't have the list in there just yet (meaning the header keys is `nil`), we'll fall back to the current list, which we won't remove until we do a major version bump. We could print a warning about this. 

## Conclusion

Both implementations seem alike, because they’re pretty much the same. The biggest difference is that the explicit solution moves the headers to the users’ configurations and makes the default list shorter. In both solutions, we’ll have a fallback list, and every edit is a complete overwrite of that default.

*Note*: decisions made here will have effect on the Ruby side. Although it's easier to implement a solution that won't work in Elixir, we'd like to keep both implementations similar. What do you think about this for both the Ruby and Elixir implementations, @matsimitsu and @tombruijn?

### PR DESCRIPTION
**Pull request by jeffkreeftmeijer**
Adds an explicit header whitelist to the configuration file or environment variables created/suggested by the installer. Closes #317.

## TODO
- [x] Add a warning if no request headers are configured (https://github.com/appsignal/appsignal-elixir/pull/327#discussion_r180010430 )

*Note*: This is a work in progress until we've decided on the final list of default headers.

### CODE REVIEW

**Comment by roytomeij** [comment]
💯 What do we need to do to ship this, and can we have it in the Ruby gem at the same time?

**Comment by tombruijn** [comment]
Related Ruby PR: https://github.com/appsignal/appsignal-ruby/pull/406

**Comment by tombruijn** [comment]
@jeffkreeftmeijer updated the header list. Now ready for review?

**Comment by jeffkreeftmeijer** [comment]
Yes! PLease do. :)

File: lib/appsignal/config.ex [hunk 1]
```diff
@@ -36,13 +42,32 @@ defmodule Appsignal.Config do
...
+      #{multiline_suggested_request_headers()}
+        )
+
+      Or set the APPSIGNAL_REQUEST_HEADERS environment variable:
+
+        $ export APPSIGNAL_REQUEST_HEADERS="#{single_line_suggested_request_headers()}"
+
+      Please check https://github.com/appsignal/appsignal-elixir/pull/336 for more information on this change.
+      """)
+    end
```
Comment (tombruijn) [inline lib/appsignal/config.ex]: This warning is also being printed during the installation now, which is kind of weird. I think it's missing the config for the installation task when it sends a demo sample.

**Comment by tombruijn** [comment]
Ah I can confirm it also still happens with the proposed fix if you run the installer again in a project with a config already. Less of a problem and I would suggest it's a non-fix because you wouldn't run the installer again if you've run it already.

File: lib/appsignal/config.ex [hunk 2]
```diff
@@ -232,6 +257,24 @@ defmodule Appsignal.Config do
...
+  end
+
+  def multiline_suggested_request_headers do
+    Enum.map_join(@suggested_request_headers, "\n", fn(row) ->
+      "    #{Enum.join(row, " ")}"
+    end)
+  end
+
+  defp test? do
+    Mix.env
```
Comment (doughsay) [inline lib/appsignal/config.ex]: @jeffkreeftmeijer It seems you've added a runtime dependency to Mix here, this breaks our distillery based releases which don't have Mix at runtime...  Just upgraded from 1.5.0 to 1.6.0 and our app no longer boots.

Comment (tombruijn) [reply lib/appsignal/config.ex]: Hi @doughsay , Sorry about that. Please downgrade to 1.5.0 until we've released the fix in PR https://github.com/appsignal/appsignal-elixir/pull/346 , which will hopefully be tomorrow!

Comment (doughsay) [reply lib/appsignal/config.ex]: thanks! ;)

Comment (lucasmazza) [reply lib/appsignal/config.ex]: Great job on fixing this so fast 🎉 

We have got the same issue as @doughsay with our distillery based deploys, so subscribe me to the next release notification :)

Comment (tombruijn) [reply lib/appsignal/config.ex]: @lucasmazza @doughsay We've released the 1.6.1 release yesterday that reverses this change. Please upgrade and let us know if you're running into any other issues.

**Pull request merged**